File: python/day4.py
from functools import reduce
from math import ceil

from utils import file, read_lines, time_taken


def get_test(part: int):
    return read_lines(file(4, part))


def part1(gears: list[str]):
    return int(gears[0]) * 2025 // int(gears[-1])


def part2(gears: list[str]):
    return ceil(int(gears[-1]) * 10_000_000_000_000 / int(gears[0]))


def part3(gears: list[str]):
    start = int(gears[0])
    end = int(gears[-1])
    end_teeth = reduce(
        lambda acc, line: acc * line[-1] // line[0],
        map(lambda line: list(map(int, line.split("|"))), gears[1:-1]),
        start * 100,
    )

    return end_teeth // end


if __name__ == "__main__":
    print("Part 1:")
    time_taken(lambda: part1(get_test(1)))
    print("Part 2:")
    time_taken(lambda: part2(get_test(2)))
    print("Part 3:")
    time_taken(lambda: part3(get_test(3)))

File: rust/src/day_4.rs
use crate::utils::Solution;
use std::fmt::Display;

pub struct Day4;

impl Solution for Day4 {
    const DAY: u32 = 4;

    fn part1(input: &str) -> impl Display {
        let mut gears = input.lines();
        let first: u64 = gears.next().unwrap().parse().unwrap();
        let last: u64 = gears.last().unwrap().parse().unwrap();

        2025 * first / last
    }

    fn part2(input: &str) -> impl Display {
        let mut gears = input.lines();
        let first: f64 = gears.next().unwrap().parse().unwrap();
        let last: f64 = gears.last().unwrap().parse().unwrap();

        (10_000_000_000_000_f64 * last / first).ceil()
    }

    fn part3(input: &str) -> impl Display {
        let mut gears = input.lines().peekable();
        let start: u64 = gears.next().unwrap().parse().unwrap();
        let mut mid: u64 = 1;
        let mut last: u64 = 1;
        while let Some(gear) = gears.next() {
            if gears.peek().is_none() {
                last = gear.parse().unwrap();
                break;
            }
            let (a, b) = gear.split_once("|").unwrap();
            let a: u64 = a.parse().unwrap();
            let b: u64 = b.parse().unwrap();
            mid *= b / a;
        }

        100 * start * mid / last
    }
}
